connectors/sync_job_runner.py

```diff
@@ -0,0 +1,166 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License 2.0;
+# you may not use this file except in compliance with the Elastic License 2.0.
+#
+import asyncio
+import time
+
+from connectors.byoc import JobStatus
+from connectors.es import Mappings
+from connectors.filtering.validation import (
+    FilteringValidationState,
+    InvalidFilteringError,
+    validate_filtering,
+)
+from connectors.logger import logger
+
+
+class SyncJobRunningError(Exception):
+    pass
+
+
+class SyncJobRunner:
+    """The class to run a sync job."""
+
+    def __init__(
+        self,
+        source_klass,
+        sync_job,
+        connector,
+        elastic_server,
+        bulk_options,
+    ):
+        self.source_klass = source_klass
+        self.sync_job = sync_job
+        self.job_id = self.sync_job.id
+        self.connector = connector
+        self.connector_id = self.connector.id
+        self.elastic_server = elastic_server
+        self.bulk_options = bulk_options
+        self._start_time = None
+        self.running = False
+
+    async def execute(self):
+        if self.running:
+            raise SyncJobRunningError(f"Sync job #{self.job_id} is already running.")
+
+        self.running = True
+        if not await self._claim_job():
+            logger.error(f"Unable to claim job #{self.job_id}")
+            return
+
+        try:
```

Comment:
It's super hard to read such a long block of code. Is there a way to split it into several ones?